src/Dofs/sparsity_pattern.jl

```diff
@@ -96,21 +100,78 @@ function _coupling_to_local_dof_coupling(dh::DofHandler, coupling::AbstractMatri
     return outs
 end
 
-function _create_sparsity_pattern(dh::AbstractDofHandler, ch#=::Union{ConstraintHandler, Nothing}=#, sym::Bool, keep_constrained::Bool, coupling::Union{AbstractMatrix{Bool},Nothing})
+"""
+    cross_element_coupling!(dh::DofHandler, topology::ExclusiveTopology, sym::Bool, keep_constrained::Bool, couplings::Union{AbstractVector{<:AbstractMatrix{Bool}},Nothing}, cnt::Int, I::Vector{Int}, J::Vector{Int})
+
+Mutates `I, J` to account for cross-element coupling in fields with discontinuous interpolations.
+Returns the updated value of `cnt`.
+
+Used internally for sparsity patterns with discontinuous interpolations.
+"""
+function cross_element_coupling!(dh::DofHandler, topology::ExclusiveTopology, sym::Bool, keep_constrained::Bool, couplings::Union{AbstractVector{<:AbstractMatrix{Bool}},Nothing}, cnt::Int, I::Vector{Int}, J::Vector{Int})
+    for (sdh_idx, sdh) in pairs(dh.subdofhandlers)
+        # Buffering interpolation information for type stability/better allocations
+        ip_infos = InterpolationInfo[]
+        nbasefunctions = Int[]
+        for ip in sdh.field_interpolations
+            ip_info = InterpolationInfo(ip)
+            push!(ip_infos, ip_info)
+            push!(nbasefunctions, getnbasefunctions(ip))
+        end
+        isnothing(couplings) || (coupling_sdh = couplings[sdh_idx])
+        for (cell_field_i, cell_field) in enumerate(sdh.field_names)
+            fii = ip_infos[cell_field_i]
+            # Couple element for discontinuous interpolations only
+            fii.is_discontinuous || continue
+            for cell_idx in eachindex(getcells(dh.grid))
+                # Making sure current field is defined for current cell
+                cell_field ∈ dh.subdofhandlers[dh.cell_to_subdofhandler[cell_idx]].field_names || continue
```

Comment:
You can/should loop directly over the cellset associated with this subdofhandler:
```suggestion
            for cell_idx in BitSet(sdh.cellset)
```

Reply:
I think we can do it but not sure if should (would require more changes for coupling different fields on different subdofhandlers), the idea here is iterating over all cells that have a field.
For example, the test:
```julia
    dh = DofHandler(grid)
    sdh1 = SubDofHandler(dh, Set(1))
    add!(sdh1, :u, DiscontinuousLagrange{RefQuadrilateral,1}()^2)
    add!(sdh1, :y, DiscontinuousLagrange{RefQuadrilateral,1}())
    add!(sdh1, :p, Lagrange{RefQuadrilateral,1}())
    sdh2 = SubDofHandler(dh, Set(2))
    add!(sdh2, :u, DiscontinuousLagrange{RefQuadrilateral,1}()^2)
    close!(dh)
```
 Would fail because the second subdofhandler doesn't have `:y` field, thus it won't couple `:u` with `:y` for second cell dofs
 would result in:
 ```julia
 ⎡⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎤
⎢⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎥
⎢⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎥
⎢⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⎥
⎢⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⎥
⎣⣿⣿⣿⣿⠀⠀⠀⠀⣿⣿⣿⣿⎦
```
while the current implementation results in:
```julia
⎡⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎤
⎢⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎥
⎢⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⎥
⎢⣿⣿⣿⣿⣿⣿⣿⣿⠀⠀⠀⠀⎥
⎢⣿⣿⣿⣿⣿⣿⠀⠀⣿⣿⣿⣿⎥
⎣⣿⣿⣿⣿⣿⣿⠀⠀⣿⣿⣿⣿⎦
```
Notice some couplings are missing.
Please correct me if I'm wrong/they shouldn't be coupled.

Reply:
I am not sure I follow, sorry, can you post the full code for producing those two patterns so I can play around a bit? Is the problem that you only add the dofs in one "direction" when looping over the cells?

Reply:
Yes, the problem is that the code added dofs in one direction, so if one cell has more fields than its neighbor they will be coupled in one direction only. This was mitigated by:
https://github.com/Ferrite-FEM/Ferrite.jl/blob/bb16acbee3f77d283a5864788dc54cdf75c4b863/src/Dofs/sparsity_pattern.jl#L110-L122

https://github.com/Ferrite-FEM/Ferrite.jl/blob/bb16acbee3f77d283a5864788dc54cdf75c4b863/src/Dofs/sparsity_pattern.jl#L161-L164

In the last commits, so we add once in each direction.
For the full code IIRC:
```julia
grid = generate_grid(Quadrilateral, (2, 1))
topology = ExclusiveTopology(grid)

dh = DofHandler(grid)
sdh1 = SubDofHandler(dh, Set(1))
add!(sdh1, :u, DiscontinuousLagrange{RefQuadrilateral,1}()^2)
add!(sdh1, :y, DiscontinuousLagrange{RefQuadrilateral,1}())
add!(sdh1, :p, Lagrange{RefQuadrilateral,1}())
sdh2 = SubDofHandler(dh, Set(2))
add!(sdh2, :u, DiscontinuousLagrange{RefQuadrilateral,1}()^2)
close!(dh)

create_sparsity_pattern(dh; topology = topology, elements_coupling = trues(3,3))
```

Reply:
Okay, so right now we add in both directions for every cell then?

Reply:
indeed